### PR TITLE
remove actions-rs

### PR DESCRIPTION
**Pull request by Emilgardis**
This essentially removes all actions-rs actions and replaces it with what is needed.

Reason for this is due to actions-rs being seemingly unmaintained

### CODE REVIEW

**Comment by Emilgardis** [comment]
this follows the general consensus as discussed in https://github.com/actions-rs/toolchain/issues/216

it also adds a matcher, which would otherwise be gone

File: .github/actions/cargo-install-upload-artifacts/action.yml [hunk 1]
```diff
@@ -29,17 +29,15 @@ runs:
...
-        args: >
+      run: >
+        cargo install
           --locked
           --path .
           --target ${{ inputs.target }}
           --all-features
           --root ${{ steps.metadata.outputs.out-dir }}
           --bins
-        use-cross: true
```
Comment (Emilgardis) [inline .github/actions/cargo-install-upload-artifacts/action.yml:42]: the use-cross is technically a bug, cross does not currently support install

Comment (Emilgardis) [reply .github/actions/cargo-install-upload-artifacts/action.yml:42]: also, I love the ouroboro implication :D

File: .github/actions/setup-rust/action.yml [hunk 2]
```diff
@@ -15,17 +15,15 @@ inputs:
 runs:
   using: composite
   steps:
-    # Work around for sporadic issues with rustup self-update on windows
-    - run: rustup set auto-self-update disable
-      if: contains(runner.os, 'windows')
-      shell: bash
```
Comment (Emilgardis) [inline .github/actions/setup-rust/action.yml:21]: dtolnay/rustc-toolchain implements this, always.

File: .github/actions/setup-rust/rust.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+                    "severity": 1,
+                    "message": 4,
+                    "code": 3
+                },
+                {
+                    "regexp": "^([\\s->=]*(.*):(\\d*):(\\d*)|.*)$",
+                    "file": 2,
+                    "line": 3,
+                    "column": 4
+                }
```
Comment (Emilgardis) [inline .github/actions/setup-rust/rust.json:17]:
```suggestion
                {
                    "regexp": "^([\\s->=]*(.*):(\\d*):(\\d*)|.*)$",
                    "file": 2,
                    "line": 3,
                    "column": 4
                },
                {
                    "regexp": "^\\d* *(\\||=)(.*)$",
                    "message": 0,
                    "loop": true
                }
```

I wanted it to be this, but it's not possible due to https://github.com/actions/toolkit/issues/319 (which should probably be re-opened or re-issued)

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -58,11 +55,8 @@ jobs:
           toolchain: 1.58.1 # MSRV, Minimally Supported Rust Version. Make sure to update README.md
 
       - name: Run clippy
-        uses: actions-rs/cargo@v1
-        with:
-          command: clippy
-          args: --locked --all-targets --all-features --workspace -- -D warnings
-
+        run: cargo clippy --locked --all-targets --all-features --workspace -- -D warnings
+        shell: bash
```
Comment (Emilgardis) [inline .github/workflows/ci.yml:59]: not needed
```suggestion
```

File: .github/actions/setup-rust/rust.json (hunk 3 above)
Comment (Emilgardis) [reply .github/actions/setup-rust/rust.json:17]: technically, this would be possible if it was possible for cargo to output json diagnostics and human diagnostics, the error annotations in the log would be strange however.

File: .github/actions/setup-rust/rust.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            "pattern": [
+                {
+                    "regexp": "^\\[cross\\] (warning|error): (.*)$",
+                    "severity": 1,
+                    "message": 2
+                }
+            ]
+        }
+    ]
+}
```
Comment (Alexhuszagh) [inline .github/actions/setup-rust/rust.json:31]: This needs a newline.

**Review by Alexhuszagh — APPROVED** [review]
lgtm, just needs a single newline in the JSON file.

**Comment by Alexhuszagh** [comment]
Weird, the changelog check only uses `serde_json` here, so I wonder. Also, the filter shouldn't be running on that directory.
https://github.com/cross-rs/cross/runs/7379312457?check_suite_focus=true#step:3:71

**Comment by Emilgardis** [comment]
bors r=@Alexhuszagh

**Comment by Emilgardis** [comment]
bors try --target x86_64-unknown-linux-gnu

will squash (and rebase) on success)

**Comment by Alexhuszagh** [comment]
bors try --target x86_64-unknown-linux-gnu

going to give it a careful re-review, but looks good so far.

**Review by reitermarkus — APPROVED** [review]

**Comment by reitermarkus** [comment]
Relatedly, but not necessarily relevant for this particular PR, I have been thinking about creating a `cross-rs/install` action which would install the `cross` binary from our GitHub releases rather than having to build it from source in CI every time.

**Comment by Emilgardis** [comment]
> Relatedly, but not necessarily relevant for this particular PR, I have been thinking about creating a `cross-rs/install` action which would install the `cross` binary from our GitHub releases rather than having to build it from source in CI every time.

That sounds good, it is possible to use https://github.com/taiki-e/install-action for now, but would be neat to have, and easy to support 👍🏼

**Comment by reitermarkus** [comment]
Neat, didn't know about https://github.com/taiki-e/install-action.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -199,7 +188,6 @@ jobs:
         if: matrix.deploy
         with:
           target: ${{ matrix.target }}
-          image: ${{ steps.build-docker-image.outputs.image }}
```
Comment (Alexhuszagh) [inline .github/workflows/ci.yml:202]: Just confirming that this is never used, correct? We don't use `image` prior, just `target`, as `${{ inputs.target }}`?

**Comment by Alexhuszagh** [comment]
Still looks good after #1040, am I free to merge this @Emilgardis?

File: .github/workflows/ci.yml (hunk 6 above)
Comment (Emilgardis) [reply .github/workflows/ci.yml:202]: yes, the only input is target: https://github.com/Emilgardis/cross/blob/eeae4f6ead3ce44cb687b126541e70bc7649e1a9/.github/actions/cargo-install-upload-artifacts/action.yml#L3-L6

**Comment by Emilgardis** [comment]
Yep @Alexhuszagh

**Comment by Alexhuszagh** [comment]
bors r=Alexhuszagh,reitermarkus

**Comment by Emilgardis** [comment]
bors try --target *windows-msvc

File: .github/actions/cargo-install-upload-artifacts/action.yml [hunk 7]
```diff
@@ -29,17 +29,15 @@ runs:
...
+      run: |
+        cargo install \
+          --locked \
+          --path . \
+          --target ${{ inputs.target }} \
+          --all-features \
+          --root ${{ steps.metadata.outputs.out-dir }} \
           --bins
-        use-cross: true
+      shell: bash
```
Comment (Alexhuszagh) [inline .github/actions/cargo-install-upload-artifacts/action.yml:40]: I think this is the issue, actually. This is causing the following issue in our actions, as shown [here](https://github.com/cross-rs/cross/actions/runs/3168710823/jobs/5160122507#step:13:57):

It's installing `cross` to `C:UsersRUNNER~1AppDataLocalTemptmp.FKQJyKb3ZP\bin\cross.exe`, `outdir` is `/tmp/tmp.FKQJyKb3ZP`. In previous [runs](https://github.com/cross-rs/cross/actions/runs/3164781446/jobs/5153342326#step:13:57), `cross` wwas installed to `C:\Users\RUNNER~1\AppData\Local\Temp\tmp.2K6uluIXkp\bin\cross.exe` and had the outdir be `/tmp/tmp.2K6uluIXkp`.

In short, `shell: bash` is undoing the `\` characters for our paths.

File: .github/actions/cargo-publish/action.yml [hunk 8]
```diff
@@ -28,10 +28,8 @@ runs:
 
     - name: Log into crates.io
       if: github.event_name != 'pull_request'
-      uses: actions-rs/cargo@v1
-      with:
-        command: login
-        args: -- ${{ inputs.cargo-registry-token }}
+      run: cargo login -- ${{ inputs.cargo-registry-token }}
+      shell: bash
```
Comment (Alexhuszagh) [inline .github/actions/cargo-publish/action.yml:32]: This likely doesn't need a shell, does it?

File: .github/actions/cargo-publish/action.yml [hunk 9]
```diff
@@ -57,7 +55,5 @@ runs:
           ${{ steps.download-artifacts.outputs.download-path }}/cross-*/*
 
     - name: Publish crate
-      uses: actions-rs/cargo@v1
-      with:
-        command: publish
-        args: ${{ !startsWith(github.ref, 'refs/tags/v') && '--dry-run' || '' }}
+      run: cargo publish ${{ !startsWith(github.ref, 'refs/tags/v') && '--dry-run' || '' }}
+      shell: bash
```
Comment (Alexhuszagh) [inline .github/actions/cargo-publish/action.yml:59]: Ditto.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -94,7 +84,6 @@ jobs:
       - uses: ./.github/actions/setup-rust
       - run: cargo xtask ci-job check
         id: check
-        shell: bash
```
Comment (Alexhuszagh) [inline .github/workflows/ci.yml:97]: Ditto.

File: .github/actions/setup-rust/action.yml [hunk 11]
```diff
@@ -15,17 +15,15 @@ inputs:
...
       with:
         toolchain: ${{ inputs.toolchain }}
         target: ${{ inputs.target }}
-        default: true
         components: ${{ inputs.components }}
-        profile: minimal
+
+    - name: Install rust matcher
+      run: echo "::add-matcher::.github/actions/setup-rust/rust.json"
+      shell: bash
```
Comment (Alexhuszagh) [inline .github/actions/setup-rust/action.yml:27]: Due to the echo, this does need a shell.

**Review by Alexhuszagh — CHANGES_REQUESTED** [review]
bors try-
Canceled the workflow because the issue was detected and wasn't a spurious error. See the comments above for removing `shell: bash` to fix the new CI actions and workflows.

**Comment by Emilgardis** [comment]
bors try --target *windows-msvc

File: .github/workflows/ci.yml (hunk 10 above)
Comment (Emilgardis) [reply .github/workflows/ci.yml:97]: this has been removed already

**Comment by Emilgardis** [comment]
bors try --target *windows-msvc

**Comment by Emilgardis** [comment]
bors try-
bors try --target *windows-msvc

**Comment by Emilgardis** [comment]
bors try-
bors try --target *windows-msvc

**Comment by Emilgardis** [comment]
bors try --target *windows-msvc

**Comment by Emilgardis** [comment]
it works!

image

**Review by Alexhuszagh — APPROVED** [review]
bors r=Alexhuszagh,reitermarkus
